lib/trino-filesystem-azure/src/main/java/io/trino/filesystem/azure/AzureOutputStream.java

```diff
@@ -135,32 +129,106 @@ private void ensureOpen()
         }
     }
 
-    @Override
-    public void close()
-            throws IOException
+    private void ensureCapacity(int extra)
     {
-        if (!closed) {
-            closed = true;
-            try {
-                stream.close();
+        int capacity = min(writeBlockSizeBytes, bufferSize + extra);
+        if (buffer.length < capacity) {
+            int target = max(buffer.length, initialBufferSize);
+            if (target < capacity) {
+                target += target / 2; // increase 50%
+                target = clamp(target, capacity, writeBlockSizeBytes);
             }
-            catch (IOException e) {
-                // Azure close sometimes rethrows IOExceptions from worker threads, so the
-                // stack traces are disconnected from this call. Wrapping here solves that problem.
-                throw new IOException("Error closing file: " + location, e);
+            buffer = Arrays.copyOf(buffer, target);
+            memoryContext.setBytes(buffer.length);
+        }
+    }
+
+    private void flushBuffer(boolean finished)
+            throws IOException
+    {
+        // skip multipart upload if there would only be one staged block
+        if (finished && !multiStagingStarted) {
+            BlobOutputStream blobOutputStream = blockClient
+                    .getBlobOutputStream(overwrite);
+
+            blobOutputStream.write(buffer, 0, bufferSize);
+            blobOutputStream.close();
+            return;
+        }
+
+        if ((bufferSize == writeBlockSizeBytes) || (finished && (bufferSize > 0))) {
```

Comment:
I'd invert this and exit early to reduce the extra level of intent here

Reply:
This is the same as in the S3OutputStream. I can follow up with the change there